Add the flow for new joining nodes [WIP]

fizyk20 opened the pull request:
This adds the flow for handling nodes that are trying to join the network for the first time.

This PR is WIP; for now, the focus is just modifying the graphs. An implementation and tests will be added in the future.

jeanphilippeD commented:
Link to rendered view:
https://raw.githack.com/fizyk20/routing_model/joining-flow/index.html

jeanphilippeD commented on index.html:
```diff
@@ -889,10 +928,10 @@ <h2>Relocating a member of our section away from it</h2>
+				There are 4 states the node may be in: State::InitialRelocation, State::RelocatingAgeIncrease,
+				State::RelocatingHop, and State::RelocatingBackOnline.<br />
+                Nodes will be selected in this order: State::RelocatingAgeIncrease, State::RelocatingHop,
+				State::InitialRelocation and then State::RelocatingBackOnline and tie break by age then name.<br />
```
I think I'd put State::InitialRelocation as the last to take on since the other ones are already older.

jeanphilippeD commented:
Looks good in principle though I do have some concerns:

1- We are adding the node to our data structure without rate limiting. That could be an issue.
 I guess we could allow only 1 (or N) initially relocating node.

2- We are adding a node we can communicate with through a proxy, it might be ok since we kind of doing something similar for candidate, but we may have to be careful about it.

3-I think I was careful to avoid allowing avenue for Parsec Spam. This is the reason CandidateInfo is not voted into parsec until we know it is a valid candidate (also requiring node to send that RPC again until it get there).
 Not sure what we can do for that one.

I think we probably would want to address these (or explain them away).

fizyk20 commented:
I addressed the code comment and point 1, but I'm not sure what we can do about the other two points you had.

2 - Should we connect to multiple proxies and only get accepted as `InitialRelocation` if a supermajority accepts us? Not sure if that doesn't open a whole other can of worms.
3 - I'm not sure which part you're referring to here, but if I understand correctly, maybe my answer to point 2 would help here as well?

jeanphilippeD commented:
> 2- We are adding a node we can communicate with through a proxy, it might be ok since we kind of doing something similar for candidate, but we may have to be careful about it.
> 2 - Should we connect to multiple proxies and only get accepted as `InitialRelocation` if a supermajority accepts us? Not sure if that doesn't open a whole other can of worms.

=> I think it is not necessarily an issue, to have one of our node communicating with us through proxy rather than direct connection. But it is a quirk, and it would be great to avoid it.
=> Maybe a way to avoid it is establishing a direct connection as we add it? 

>3-I think I was careful to avoid allowing avenue for Parsec Spam. This is the reason CandidateInfo is not voted into parsec until we know it is a valid candidate (also requiring node to send that RPC again until it get there).
Not sure what we can do for that one.
> 3 - I'm not sure which part you're referring to here, but if I understand correctly, maybe my answer to point 2 would help here as well?

I don't think 2 would help so much here, we still have cheap single node getting us to vote something in PARSEC. But it is probably something we can deal with differently (proof of work of sort), so maybe just highlighting the fact is enough.

jeanphilippeD commented on index.html:
```diff
@@ -1443,6 +1482,66 @@ <h2>Handling members of our section going online or offline</h2>
+		RPC --"Rpc::RequestRelocate"--> Check
+		Check(("Check"))
+		
+		Check --"can_accept"--> VoteNewNode
```
We can have that check here as well, to avoid Parsec Spam, but we need to check on consensus, otherwise we will accept more nodes that we want.
